### PR TITLE
ttl: convert DistSQLPlanner span keys to nearby row keys

### LINKED ISSUE #102348: ttl: convert DistSQLPlanner span keys to nearby row keys
Using the DistSQLPlanner span keys directly can lead to errors like this
```
error decoding EncDatum of <x>: did not find terminator 0x0 in buffer <x>
```
when the key is truncated in the middle of a datum.

To fix this, the TTL processor bounds start should be the first row `span.Key <=` and the bounds end should be the first row `<= span.EndKey`.

Jira issue: CRDB-27405

Epic CRDB-18322

### PR DESCRIPTION
**Pull request by ecwall**
Fixes #102348

Previously the TTL DistSQL processors used the DistSQL planner span keys
directly to determine the bounds for their SELECT queries. This lead to errors
like
```
error decoding EncDatum of <x>: did not find terminator 0x0 in buffer <x>
```
when the key was truncated in the middle of a datum.

This change prevents the decoding error by using the nearest row from KV:
`span.Key` -> KV scan to find the closest row `>= span.Key`
`span.EndKey` -> KV reverse scan to find the closest row `< span.EndKey`

Release note (bug fix): Previously, row level TTL jobs could incorrectly
process a table that spanned multiple ranges in rare cases. The bug has been
present since 22.2.0 and is now fixed. You were affected by this bug if you
saw error messages like
"error decoding EncDatum of ...: did not find terminator ... in buffer ...".

### CODE REVIEW

**Review by rharding6373 — APPROVED** [review]
`SpanToQueryBounds` :lgtm: My only question is for curiosity's sake.

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102324#-:-NU2bTJVEBlJps8B72s8:b-suuqbb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ecwall, @rafiss, and @yuzefovich)

___
*[`pkg/sql/ttl/ttlbase/ttl_helpers.go` line 116 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/102324#-NU2cewR5dusAaqZ3pFh:-NU2cewR5dusAaqZ3pFi:b-jy6hnp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5da933d923079af08d1b8effb3c8dba53efa87f2/pkg/sql/ttl/ttlbase/ttl_helpers.go#L116)):*
> ```Go
> 		endPlaceholderOffset,
> 		endKeyCompareOps,
> 		true, /*inclusive*/
> ```

Could you explain why this change was needed? I don't have a solid understanding of TTL jobs...
___
*[`pkg/sql/ttl/ttljob/ttljob_processor_test.go` line 32 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/102324#-NU2bcZ609CEcMunJRBU:-NU2bcZ609CEcMunJRBV:b-scsjab) ([raw file](https://github.com/cockroachdb/cockroach/blob/5da933d923079af08d1b8effb3c8dba53efa87f2/pkg/sql/ttl/ttljob/ttljob_processor_test.go#L32)):*
> ```Go
> )
> 
> func TestSpanToQueryBounds(t *testing.T) {
> ```

I like this unit test, I think it's very thorough.

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102324)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss, @rharding6373, and @yuzefovich)

___
*[`pkg/sql/ttl/ttlbase/ttl_helpers.go` line 116 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/102324#-NU2cewR5dusAaqZ3pFh:-NU2h9A2A0LS3dHtoynU:b-3qguw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5da933d923079af08d1b8effb3c8dba53efa87f2/pkg/sql/ttl/ttlbase/ttl_helpers.go#L116)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Could you explain why this change was needed? I don't have a solid understanding of TTL jobs...
</blockquote></details>

I'll add a comment for this.

This was originally exclusive because `QueryBounds.End` was based on the DistSQLPlanner `Span.EndKey` which is exclusive so we don't want to select it when processing a span (if that key needs to be selected, it will end up being the next span's start key and it will be selected there).

This PR changes `QueryBounds.End` to instead be based on the first row `< Span.EndKey`. In order to make sure that first row is selected, `QueryBounds.End` needs to be inclusive instead of exclusive (ttljob_test.go has some test cases that prove this by failing if this line isn't changed).

In other words, we went from `[1, 2, 3, 4)` to `[1, 2, 3]`.

**Comment by ecwall** [comment]
bors r=rharding6373

**Review by yuzefovich — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 5 of 5 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102324#-:-NU36UtC2NfMM7VY8L88:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ecwall, @rafiss, and @rharding6373)

___
*[`-- commits` line 33 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102324#-NU31fkAEAnAzQajya-V):*
nit: I would provide more details in the release note. "Partial keys" (at least to me) suggests it's something about the partial indexes, or some keys not being fully specified (that's why they are "partial"), but neither is true. The bug occurs due to some internal CRDB details (range splits at unlucky positions within a single table), "spans" is also relatively low level internal detail. The audience of the release notes are the users, so the note should be written accordingly. How about something like this:
```
Release note (bug fix): Previously, CockroachDB could incorrectly process TTL requirement on
a table that spanned multiple ranges in some cases. The bug has been present since TTL feature
was introduced and is now fixed. You were affected by this bug if you saw error messages like
"error decoding EncDatum of ...: did not find terminator ...".
```
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 416 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102324#-NU35GeoCdMVeEM17Mq2:-NU35GeoCdMVeEM17Mq3:bpmg66w) ([raw file](https://github.com/cockroachdb/cockroach/blob/577c8d8a4ec208f5c284657dfa05be51a8e8514f/pkg/sql/ttl/ttljob/ttljob_processor.go#L416)):*
> ```Go
> 	}
> 	// If span has 0 keys then continue to next span.
> 	if len(endKeyValues) == 0 {
> ```

nit: isn't this case a contradiction? Should we return an assertion failed error?
___
*[`pkg/sql/ttl/ttljob/ttljob_processor_test.go` line 32 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/102324#-NU2bcZ609CEcMunJRBU:-NU36PHz1fPfslYxcAgS:bt11d19) ([raw file](https://github.com/cockroachdb/cockroach/blob/5da933d923079af08d1b8effb3c8dba53efa87f2/pkg/sql/ttl/ttljob/ttljob_processor_test.go#L32)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

I like this unit test, I think it's very thorough.
</blockquote></details>

+1, great test!

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102324)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss, @rharding6373, and @yuzefovich)

___
*[`-- commits` line 33 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102324#-NU31fkAEAnAzQajya-V):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I would provide more details in the release note. "Partial keys" (at least to me) suggests it's something about the partial indexes, or some keys not being fully specified (that's why they are "partial"), but neither is true. The bug occurs due to some internal CRDB details (range splits at unlucky positions within a single table), "spans" is also relatively low level internal detail. The audience of the release notes are the users, so the note should be written accordingly. How about something like this:
```
Release note (bug fix): Previously, CockroachDB could incorrectly process TTL requirement on
a table that spanned multiple ranges in some cases. The bug has been present since TTL feature
was introduced and is now fixed. You were affected by this bug if you saw error messages like
"error decoding EncDatum of ...: did not find terminator ...".
```
</blockquote></details>

Good call, updated.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 416 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102324#-NU35GeoCdMVeEM17Mq2:-NU3EsNz4WfFVqaHi8Tf:b-wwpkra) ([raw file](https://github.com/cockroachdb/cockroach/blob/577c8d8a4ec208f5c284657dfa05be51a8e8514f/pkg/sql/ttl/ttljob/ttljob_processor.go#L416)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: isn't this case a contradiction? Should we return an assertion failed error?
</blockquote></details>

We talked about this in slack - I'll add a comment as well:

```
kvDB.Scan(ctx, partialStartKey, partialEndKey, maxRows)
```
and 
```
kvDB.ReverseScan(ctx, partialStartKey, partialEndKey, maxRows)
```
are non-transactional so the row could have been deleted in between the calls.

**Comment by ecwall** [comment]
bors r=rharding6373,yuzefovich
